src/query/execute_statement.c

```diff
@@ -18135,16 +18214,410 @@ get_dblink_info_from_dbserver (PARSER_CONTEXT * parser, const char *server, DB_V
 	}
 
-      error = db_make_string (&(out_val[0]), dblink_url);
+      error = db_make_string_copy (&(out_val[0]), dblink_url);
     }
 
 error_end:
   AU_ENABLE (au_save);
 
+  pr_clear_value (&pwd_val);
   while (--cnt >= 0)
     {
       pr_clear_value (&(values[cnt]));
     }
 
   return error;
 }
+
+
+#define DBLINK_PASSWORD_MAX_LENGTH      (128)
+#define DBLINK_PASSWORD_CONFUSED_LENGTH (6)	// include 4(int) + 1(unsigned char) +  1(unsigned char)
+// Valid data size is the largest multiple of 3 less than or equal to DBLINK_PASSWORD_CIPHER_LENGTH.
+#define DBLINK_PASSWORD_CIPHER_LENGTH   (DBLINK_PASSWORD_MAX_LENGTH + DBLINK_PASSWORD_CONFUSED_LENGTH)
+#define DBLINK_PASSWORD_PAD_LENGTH      (40)	// include 2(length) + 2(mk) + 2(length) + 32(mk), Must be 4 or more
+#define DBLINK_PASSWORD_MAX_BUFSIZE  ((int)(DBLINK_PASSWORD_CIPHER_LENGTH / 3 * 4) + DBLINK_PASSWORD_PAD_LENGTH)
+
+/*
+ * pt_check_dblink_password ()  : Check the validity of the entered password.
+ *
+ * return		  : NO_ERROR or error code.
+ * parser(in)		  : Parser context.
+ * passwd(in)             : Password entered by user(or parser)
+ * cipher_buf(out)	  : Encrypted password
+ * ciper_buf_size(in)	  : Size of cipher_buf
+ * 
+ * Remark: 
+ *     Checks whether the entered password is a raw password or an encrypted password,
+ *     if it is a raw password, encrypt it.
+ *     The length of the raw password is shorter than DBLINK_PASSWORD_MAX_LENGTH.
+ *     If it is an encrypted password, it must have a specific length.
+ *     Even if there is no raw password, it will be returned as an encrypted password.
+ */
+int
+pt_check_dblink_password (PARSER_CONTEXT * parser, const char *passwd, char *cipher_buf, int ciper_buf_size)
+{
+  DB_VALUE val;
+  char *str;
+  int max_len = DBLINK_PASSWORD_MAX_BUFSIZE;
+  int err = ER_FAILED;
+  int length = 0;
+
+  /* Adjust the length so that it is a multiple of 4. */
+  max_len >>= 2;
+  max_len <<= 2;
+
+  if (ciper_buf_size <= max_len)
+    {
+      return ER_FAILED;
+    }
+
+#ifndef NDEBUG
+  if (passwd)
+    {
+      if (strcmp (passwd, "test_check") == 0)
+	{
+	  ciper_func_test (parser, false);
+	  return ER_FAILED;
+	}
+      else if (strcmp (passwd, "test_time") == 0)
+	{
+	  ciper_func_test (parser, true);
+	  return ER_FAILED;
+	}
+    }
+#endif
+
+  if (passwd && *passwd)
+    {
+      length = strlen (passwd);
+    }
+
+  db_make_null (&val);
+  if (length <= DBLINK_PASSWORD_MAX_LENGTH)
+    {
+      // The raw password entered by the user.
+      err = get_dblink_password_encrypt (passwd, &val, true);
+      if (err == NO_ERROR)
+	{
+	  str = (char *) db_get_string (&val);
+	  if (!str)
+	    {
+	      err = ER_FAILED;
+	    }
+	  else
+	    {
+	      strcpy (cipher_buf, str);
+	      err = NO_ERROR;
+	    }
+	}
+    }
+  else if (length == max_len)
+    {
+      // A encrypted password from the raw password.      
+      err = get_dblink_password_decrypt (passwd, &val);
```

Comment:
If it is an encrypted password, why is it decrypted?
I'm curious as it's the opposite of encrypting the raw password.

Reply:
The length comparison only confirms that it is likely an encrypted password.
It is necessary to check whether the decryption process is successful.

Reply:
There is not 'err != NO_ERROR'.
And the decryption will proceed in do_create_server() or ...bind(), does it have to be checked here?

Reply:
Returns err if 'err != NO_ERROR'.

This function verifies the validity of the PASSWORD item read in the parsing stage.
When the user enters a manipulated value, an error can be raised without going to do_create_server().

Reply:
Does it have to be checked in parser?
Is it a problem if it is checked in do_create_server()?

Reply:
It's just a time difference, but it shouldn't be a problem.
If there is no problem in the ALTER SERVER statement, it may be a good idea to remove it because it is a duplicate decryption operation.